File: CTCI/linkedlists/remove_duplicates.js
class Node { 
    // constructor 
    constructor(element) 
    { 
        this.value = element; 
        this.next = null
    } 
} 
class LinkedList { 
    constructor() 
    { 
        this.head = null; 
        this.size = 0; 
    } 
    add(data) {
        let node = new Node(data) 
        if (!this.head) {
            this.head = node
            return}
        if (!this.head.next){
            this.head.next = node 
            return
        } 
        // otherwise find the tail 
        let current = this.head 
        while (current.next) {
                current = current.next
        }
        current.next = node
        return
    }
    print() {
        let current = this.head 
        while (current) {
            console.log(current.value)
            current = current.next
        }
    }

} 
let l1 = new LinkedList()
l1.add(1)
l1.add(2)
l1.add(3)
l1.add(2)
l1.add(2)
l1.add(5)

function remove_duplicates(list) {
    let hash = {} 
    let head = list.head 
    let prev = list.head 
    if (!head.next) return list 
    while (head) {
        if (!hash[head.value]) {
            hash[head.value] = 1
            prev = head 
        }
        else {
            let next = head.next 
            prev.next = next  
        }
        head = head.next
        
    }
    return list 
}

console.log(remove_duplicates(l1))
l1.print()

File: CTCI/linkedlists/remove_duplicates.py
# 1 2 33 44 -> 1 2 
class ListNode:
    def __init__(self, val = 0, next = None):
        self.val = val 
        self.next = next 
n1 = ListNode(1)
n2 = ListNode(2)
n1.next = n2 
n3 = ListNode(2)
n4 = ListNode(3)
n5 = ListNode(3)
n6 = ListNode(4)

n3.next = n4 
n4.next = n5 
n5.next = n6
n2.next = n3 # 1224 -> 14 

def remove_duplicates(head):
    result = slow = ListNode(0, next =head)
    while head:
        if head.next and head.val == head.next.val:
            while head.next and head.val == head.next.val:
                print("head:",head.val)
                head = head.next 
            print("slow:",slow.val)
            slow.next = head.next 
        else:
            slow = slow.next 
        head = head.next
    return result.next
result = remove_duplicates(n1)
while (result):
    print(result.val)
    result = result.next